cpp_template: Add Python mechanisms to pave way for C++ template binding.

EricCousineau-TRI opened the pull request:
Related to exposing templates in #7660.

Final PR is to combine with #7732 and incorporate C++ glue:
https://github.com/EricCousineau-TRI/repro/blob/fa4e6a0/python/bindings/pymodule/tpl/cpp_template.h#L45

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7757)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
+@jadecastro for feature review, please (if it's not too much).
+@jwnimmer-tri for platform review, please.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-:-L2i1bIswmMpszz5zNcS:be0tlw2)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Pass complete.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[bindings/pydrake/util/cpp_template.py, line 122 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3-Rcc8Ifyuyef-Cs57:-L3-Rcc8Ifyuyef-Cs58:bmnh6ka) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L122)):*
> ```Python
>                 # Assume scalar.
>                 param = (param,)
>             else:
> ```

BTW, I don't think this branch is checked in the unit tests?

---

*[bindings/pydrake/util/cpp_template.py, line 138 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3-K6RC8jDc8XbCKhSC:-L3-K6RC8jDc8XbCKhSD:b-rzxqdu) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L138)):*
> ```Python
> 
>     def _on_add(self, param, instantiation):
>         # To be overrode by child classes.
> ```

BTW ("overrode by" -> "overridden by")

---

*[bindings/pydrake/util/cpp_template.py, line 159 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3-L-WEBl3PjaR3T6ID:-L3-L-WEBl3PjaR3T6IE:bwnikq6) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L159)):*
> ```Python
> 
> 
> class TemplateMethod(TemplateFunction):
> ```

BTW, can this just derive directly from the `Template` class, removing `TemplateFunction`, or is there some intended use case I'm missing? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-:-L3-zBB2Duh-IT5NC0Oc:bilrup7)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3012_WAlU6w195MVvF:-L3012_WAlU6w195MVvG:b-a17o8x) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/test/cpp_template_test.py#L10)):*
> ```Python
> 
> 
> class A(object):
> ```

BTW, please label these "dummy" classes and functions here and below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 3 files reviewed at latest revision, 14 unresolved discussions.

---

*[bindings/pydrake/util/cpp_template.py, line 9 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zfjjPm7C8CHbogBu3:-L30HKRHc1kAzPci0STT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Should this be a TODO instead?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 14 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zfr31UdaSsmK9RMoA:-L30GtuD8HDva2Bz-tSI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L14)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I was confused because I didn't initially understand that this was a stub.  Needs a comment like the 7732 above.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2k6ml7_Qe_iqnBZu5O:-L30GvBr9QacBzyoK7X6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L25)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI _Meta_ I think the `"""` nominally belongs on the following line?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zgCg6RL4PcVKF5SuR:-L30GvYVT7Qbl4DZD33I:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L25)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This appears to be a public API (no leading `_`), but I don't think its clear what the three arguments denote (or where args and kwargs forwarding goes).
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zgQgc9Mxf_CAGkRed:-L30HJPbax1R397Nqkoe:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Disallowed abbreviation.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zXGfpOlyx2pJid_SS:-L30HI7BzXjh_KYgrhXX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L37)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Should this base class ever be used directly at call sites, or is it merely convenient implementation reuse?  If the latter, consider a rename to `TemplateBase` to make that clear.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zfH2Y7BD0qOa9C0g5:-L30HLsurTgSm90iqWA0:b2ojlzy) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L56)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am not sure what this line means, given that `param` is a single item and we're not using `*args`.  (At least, I couldn't find a test case that matches this signature.)
</blockquote></details>

Done. Removed `param_default`, simplified to `allow_default`, with comments indicating what this means.

---

*[bindings/pydrake/util/cpp_template.py, line 89 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zn6fk9JDi1nWT1Sbd:-L30HPrOdyekYqCntnnG:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L89)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Ah, I see what this is doing now.  The default isn't a so much a default (that sticks around, and has a meaning on its own), its a special value that means "not yet set" ala C++ optional.  Consider if a different name would help clarify.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2znl0hU6fHDWh-DXbS:-L30HQdcCiGhEo00HXg0:b48wmxz) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L95)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I don't understand why `param_list` is given a default.
</blockquote></details>

Done. Oops! From an old version.

---

*[bindings/pydrake/util/cpp_template.py, line 117 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zo4JUYfS1bVhv7BWs:-L30HScxVH3adCLan1CT:bfmz5yd) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L117)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Should this allow `self._param_default` to be the magic `_PARAM_DEFAULT` value?  That seems like not what the resolver should be doing?  Or if so, needs some more documentation.
</blockquote></details>

Done. Removed `_param_default`.

---

*[bindings/pydrake/util/cpp_template.py, line 122 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3-Rcc8Ifyuyef-Cs57:-L30HV7HYNAjvmYbnvHh:benujs7) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L122)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, I don't think this branch is checked in the unit tests?
</blockquote></details>

Done. Added.

---

*[bindings/pydrake/util/cpp_template.py, line 138 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3-K6RC8jDc8XbCKhSC:-L30HW1TfcodFJcirl1W:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L138)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW ("overrode by" -> "overridden by")
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 159 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3-L-WEBl3PjaR3T6ID:-L30HWe5tmtg7BKyfRPx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L159)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, can this just derive directly from the `Template` class, removing `TemplateFunction`, or is there some intended use case I'm missing? 
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L3012_WAlU6w195MVvF:-L30HYdfMlklLG99f2sE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/test/cpp_template_test.py#L10)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, please label these "dummy" classes and functions here and below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 2 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/util/cpp_template.py, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zfH2Y7BD0qOa9C0g5:-L30JECpviO0dyNq0jyB:b-vptkpy) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L56)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done. Removed `param_default`, simplified to `allow_default`, with comments indicating what this means.
</blockquote></details>

OK As it turns out, what I was missing here on first read is that in Python `__getitem__` is passed a tuple if you do `template[a, b]` from calling code -- that it never receives multiple `*args`.  TIL.

---

*[bindings/pydrake/util/cpp_template.py, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30IjvDMPfGg8u46smJ:-L30IjvEXexgv4zr4-7r:b3ddjh6) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/cpp_template.py#L23)):*
> ```Python
> 
> def get_or_init(scope, name, template_cls, *args, **kwargs):
>     """
> ```

BTW Annoying PEP-257 suggests that the `"""` be immediately followed by the opening sentence, rather than a blank line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/util/cpp_template.py, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30Xh1TUyJQQREou6J1:-L30Xh1TUyJQQREou6J2:b-12raeu) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/cpp_template.py#L1)):*
> ```Python
> #!/usr/bin/env python
> ```

This doesn't have a `__main __` function; thus it should not have a shebang.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30XqE66z9TDfCRMroN:-L30XqE66z9TDfCRMroO:b-ff5o9t) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/test/cpp_template_test.py#L1)):*
> ```Python
> #!/usr/bin/env python
> ```

Python binaries compiled by Bazel should not have shebang lines.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 40 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30YQOEOsHpQ9A8s6XJ:-L30YQOEOsHpQ9A8s6XK:bs62s36) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/test/cpp_template_test.py#L40)):*
> ```Python
> class TestCppTemplate(unittest.TestCase):
>     def test_base(self):
>         tpl = m.TemplateBase("BaseTpl")
> ```

_Meta_ Some more template (confusing vs tuple) abbrevs here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-:-L30aPg9mPbrxopQB0AW:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[bindings/pydrake/util/cpp_template.py, line 56 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L2zfH2Y7BD0qOa9C0g5:-L33uxXjmG1qFyG7zL6p:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/acb159c0a2b267b94d1dd12981b08bdf502b3bd1/bindings/pydrake/util/cpp_template.py#L56)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK As it turns out, what I was missing here on first read is that in Python `__getitem__` is passed a tuple if you do `template[a, b]` from calling code -- that it never receives multiple `*args`.  TIL.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30Xh1TUyJQQREou6J1:-L33uyc057b4qvK3Enrh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/cpp_template.py#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This doesn't have a `__main __` function; thus it should not have a shebang.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_template.py, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30IjvDMPfGg8u46smJ:-L33uyC5B6ZLoZTlupOZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/cpp_template.py#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Annoying PEP-257 suggests that the `"""` be immediately followed by the opening sentence, rather than a blank line.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30XqE66z9TDfCRMroN:-L33uw78ZFt8CKxIkT8G:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/test/cpp_template_test.py#L1)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Python binaries compiled by Bazel should not have shebang lines.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 40 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30YQOEOsHpQ9A8s6XJ:-L33v4-Td4_Z9MgrjJ_5:b-c9wz07) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/test/cpp_template_test.py#L40)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

_Meta_ Some more template (confusing vs tuple) abbrevs here.
</blockquote></details>

Done.

I'd like to maintain a short suffix, as I'd like to use this with the systems framework. If you can think of another short convention (such as a trailing `_`, or just `T`), I'd be open to changing to that from `Tpl`.

Default instantiation would be `${name}`, and then template would be `${name}${suffix}`.
Options:
`Adder == AdderTpl[float]`
`Adder == Adder_[float]`
`Adder == AdderT[float]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 40 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30YQOEOsHpQ9A8s6XJ:-L33x7Di4Lvox_Szf6a1:brwpizq) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/test/cpp_template_test.py#L40)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.

I'd like to maintain a short suffix, as I'd like to use this with the systems framework. If you can think of another short convention (such as a trailing `_`, or just `T`), I'd be open to changing to that from `Tpl`.

Default instantiation would be `${name}`, and then template would be `${name}${suffix}`.
Options:
`Adder == AdderTpl[float]`
`Adder == Adder_[float]`
`Adder == AdderT[float]`
</blockquote></details>

I think for this unit test the `"Tpl"` string is just fine (I only intended my comment to cover the `tpl` variable names, but didn't write it clearly).  I think we can debate the framework's string names once we're considering that specific commit (either in PR form, or a pre-review design check).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[bindings/pydrake/util/test/cpp_template_test.py, line 40 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-L30YQOEOsHpQ9A8s6XJ:-L33xkAV4obf9HNUnoj5:b56uezr) ([raw file](https://github.com/robotlocomotion/drake/blob/5ef0a2ad0ec0d206ca8f2314c280dec4072d4da8/bindings/pydrake/util/test/cpp_template_test.py#L40)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think for this unit test the `"Tpl"` string is just fine (I only intended my comment to cover the `tpl` variable names, but didn't write it clearly).  I think we can debate the framework's string names once we're considering that specific commit (either in PR form, or a pre-review design check).
</blockquote></details>

OK Sounds good!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> !

---

Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-:-L34oO_z-WxMA2_n01jo:b593w5u)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Note: Updated since #7732 was merged.

---

Review status: 1 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757#-:-L360q8Yqfp61RIlz--b:b-64tzsr)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7757)*
<!-- Sent from Reviewable.io -->
